tests/integration/helpers.py

```diff
@@ -0,0 +1,199 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Helper utilities for integration tests."""
+
+import collections
+import inspect
+import platform
+import tarfile
+import time
+from functools import partial
+from pathlib import Path
+from string import Template
+from typing import Awaitable, Callable, ParamSpec, TypeVar, cast
+
+from pylxd import Client
+from pylxd.models.image import Image
+from pylxd.models.instance import Instance, InstanceState
+from requests_toolbelt import MultipartEncoder
+
+P = ParamSpec("P")
+R = TypeVar("R")
+S = Callable[P, R] | Callable[P, Awaitable[R]]
+
+
+async def wait_for(
+    func: S,
+    timeout: int | float = 300,
+    check_interval: int = 10,
+) -> R:
+    """Wait for function execution to become truthy.
+
+    Args:
+        func: A callback function to wait to return a truthy value.
+        timeout: Time in seconds to wait for function result to become truthy.
+        check_interval: Time in seconds to wait between ready checks.
+
+    Raises:
+        TimeoutError: if the callback function did not return a truthy value within timeout.
+
+    Returns:
+        The result of the function if any.
+    """
+    deadline = time.time() + timeout
+    is_awaitable = inspect.iscoroutinefunction(func)
+    while time.time() < deadline:
+        if is_awaitable:
+            if result := await cast(Awaitable, func()):
+                return result
+        else:
+            if result := func():
+                return cast(R, result)
+        time.sleep(check_interval)
+
+    # final check before raising TimeoutError.
+    if is_awaitable:
+        if result := await cast(Awaitable, func()):
+            return result
+    else:
+        if result := func():
+            return cast(R, result)
+    raise TimeoutError()
+
+
+IMAGE_TO_TAG = {"jammy": "22.04", "noble": "24.04"}
+
+
+def _create_metadata_tar_gz(image: str, tmp_path: Path) -> Path:
+    """Create metadata.tar.gz contents.
+
+    Args:
+        image: The ubuntu LTS image name.
+        tmp_path: Temporary dir.
+    """
+    # Create metadata.yaml
+    template = Template(
+        Path("tests/integration/testdata/metadata.yaml.tmpl").read_text(encoding="utf-8")
+    )
+    metadata_contents = template.substitute(
+        {"arch": platform.machine(), "tag": IMAGE_TO_TAG[image], "image": image}
+    )
+    meta_path = tmp_path / "metadata.yaml"
+    meta_path.write_text(metadata_contents, encoding="utf-8")
+
+    # Pack templates/ and metada.yaml
+    templates_path = Path("tests/integration/testdata/templates")
+    metadata_tar = tmp_path / Path("metadata.tar.gz")
+
+    with tarfile.open(metadata_tar, "w:gz") as tar:
+        tar.add(meta_path, arcname=meta_path.name)
+        tar.add(templates_path, arcname=templates_path.name)
+
+    return metadata_tar
+
+
+# This is a workaround until https://github.com/canonical/pylxd/pull/577 gets merged.
```

Comment:
Hey, it got merged :rocket:

Reply:
Yuup :D but it seems that it hasn't yet been released. Let me ask what the release cycle is 🚀

Reply:
It will be available coming Monday, let's see if this PR will be open until then (probably will :D )